core/src/main/java/com/linecorp/armeria/common/stream/AsyncFileWriter.java

```diff
@@ -0,0 +1,194 @@
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Maps;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * Asynchronously writes a {@code StreamMessage<HttpMessage>} to a {@link File}.
+ */
+final class AsyncFileWriter implements Subscriber<HttpData>,
+                                       CompletionHandler<Integer, Entry<ByteBuffer, ByteBuf>> {
+
+    private static final Logger logger = LoggerFactory.getLogger(AsyncFileWriter.class);
+
+    private final CompletableFuture<Void> completionFuture = new CompletableFuture<>();
+    private final Path path;
+    private final EventExecutor eventExecutor;
+    private final Set<OpenOption> options;
+    private final ExecutorService blockingTaskExecutor;
+
+    @Nullable
+    private AsynchronousFileChannel fileChannel;
+    @Nullable
+    private Subscription subscription;
+
+    // state
+    private long position;
+    private boolean writing;
+    private boolean closing;
+
+    AsyncFileWriter(StreamMessage<? extends HttpData> publisher, Path path, Set<OpenOption> options,
+                    EventExecutor eventExecutor, ExecutorService blockingTaskExecutor) {
+        this.path = path;
+        this.eventExecutor = eventExecutor;
+        this.options = options;
+        this.blockingTaskExecutor = blockingTaskExecutor;
+        publisher.subscribe(this, eventExecutor, SubscriptionOption.WITH_POOLED_OBJECTS);
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        this.subscription = subscription;
+
+        try {
+            fileChannel = AsynchronousFileChannel.open(path, options, blockingTaskExecutor);
+        } catch (IOException e) {
+            maybeCloseFileChannel(e, false);
+            return;
+        }
+
+        subscription.request(1);
+    }
+
+    @Override
+    public void onNext(HttpData httpData) {
+        if (httpData.isEmpty()) {
+            httpData.close();
+            subscription.request(1);
+        } else {
+            final ByteBuf byteBuf = httpData.byteBuf();
+            final ByteBuffer byteBuffer = byteBuf.nioBuffer();
+            writing = true;
+            try {
+                fileChannel.write(byteBuffer, position, Maps.immutableEntry(byteBuffer, byteBuf), this);
+            } catch (Throwable ex) {
+                maybeCloseFileChannel(ex, false);
+            }
+        }
+    }
+
+    @Override
+    public void onError(Throwable t) {
+        maybeCloseFileChannel(t, true);
+    }
+
+    @Override
+    public void onComplete() {
+        if (!writing) {
+            maybeCloseFileChannel(null, false);
+        } else {
+            closing = true;
+        }
+    }
+
+    CompletableFuture<Void> whenComplete() {
```

Comment:
Technically, we cannot strongly guarantee which thread the callbacks of `whenComplete()` are executed on.
If a `CompletableFuture` was completed already, the callbacks will be executed by the caller thread.
Basically, it would be good to complete a ``whenComplete()`` with `eventExecutor`.
But I don't think we need some additional operations to cover all cases.

Reply:
Understood! Thanks 👍